harness/harness-common.rb

```diff
@@ -28,3 +28,28 @@ def use_gemfile(extra_setup_cmd: nil)
   # Need to be in the appropriate directory for this...
   require "bundler/setup"
 end
+
+def get_rss
+  mem_rollup_file = "/proc/#{Process.pid}/smaps_rollup"
+  if File.exist?(mem_rollup_file)
+    # First, grab a line like "62796 kB". Checking the Linux kernel source, Rss will always be in kB.
+    rss_desc = File.read(mem_rollup_file).lines.detect { |line| line.start_with?("Rss") }.split(":", 2)[1].strip
+    1024 * rss_desc.to_i
+  else
+    # Collect our own peak mem usage as soon as reasonable after finishing the last iteration.
+    # This method is only accurate to kilobytes, but is nicely portable and doesn't require
+    # any extra gems/dependencies.
+    mem = `ps -o rss= -p #{Process.pid}`
+    1024 * mem.to_i
+  end
+end
+
+def return_results(name, values)
```

Comment:
I'm cool with it either way. The same interface can definitely work for a yjit-metrics-style JSON file: https://github.com/Shopify/yjit-metrics/compare/main...return_multiple_result_sets

I just don't know if anybody downstream is relying on easy access to the CSV files. I don't know of any, but I might not know.

Reply:
Shouldn't be a problem.

Reply:
Sounds good. Would you like the JSON conversion in this PR or a followup?

Reply:
In this PR if that's ok

Reply:
Ideally, I think we should store all the data for one benchmark run in a single JSON file (no more CSV). That way, the data for one benchmark is self-contained. We just have multiple keys in the JSON file for different types of data.

Reply:
So, without this PR we collect a CSV per harness invocation, and combine that data into one set. We actually write out a CSV (one CSV with all benchmarks), a JSON file and a text chart into the data directory. So, for any N benchmarks we run, we get three output files, named something like output_001.csv, output_001.json and output_001.txt.

We also generate N intermediate CSV files, by invoking Ruby N times. I'm pretty sure we want to keep doing that.

So mostly I think we're already doing what you want. I'll go ahead and add the ability to return multiple sets of data from any benchmark with varying names. So e.g. we could write a compile_time benchmark that ran its own trial(s) and returned the results explicitly instead of being timed by the harness, and we could write a chunky_png benchmark that could call run_benchmark 3 times as "chunky_png_no_compression", "chunky_png_good" and "chunky_png_best" (or whatever.) But all this will wind up in one JSON, one CSV and one TXT file in the data dir, just like now.

Does that sound like what you want?

Reply:
Okay, I have a reworked version: https://github.com/Shopify/yjit-bench/compare/main...compile_time_benchmark

It can return multiple benchmarks from the same benchmark file. It supports calculated benchmarks (e.g. compile time.) I've reworked the two harnesses that return results (default and warmup) to use this. I've added an example compile-time benchmark.

I'm reworking the other harnesses now, which is pretty straightforward.

Let me know if this alternate version is more what you want: https://github.com/Shopify/yjit-bench/compare/main...compile_time_benchmark

It *is* a significantly larger change. I've put it in a different branch, in case it's way off-base.

Reply:
Also, would be thrilled to pair on looking this over, if it's helpful. This is an API change, and those benefit from extra scrutiny.

Reply:
Ideally one run of a benchmark with the harness should produce one single JSON file with all the data gathered during that run. This seems to be what you have now?

Can't pair today but could tomorrow at 11 eastern or noon eastern if useful.

Reply:
Yup, that's what it does. So if you ran it with compile_time (which produces two results) and fib, you'd get two intermediate JSON files (for two harness runs), which turned into one final JSON file (for one run_benchmarks.rb run).

Noon Eastern tomorrow would be great for me. I'll send an invite.